src/Pos/Genesis.hs

```diff
@@ -109,35 +109,73 @@ stakeDistribution ts@RichPoorStakes {..} =
         else identity
     basicDist = genericReplicate sdRichmen sdRichStake ++
                 genericReplicate sdPoor sdPoorStake
-stakeDistribution (CustomStakes coins) = map (,[]) coins
-
--- | Generates genesis 'Utxo' given optional boot stakeholders list
--- and address distribution. In case boot stakeholders are supplied,
--- all the balance is distributed among them. Otherwise, txDistr is
--- set to @[]@ or left as it is (in case of 'ExplicitStakes').
-genesisUtxo :: Maybe (HashMap StakeholderId Word16) -> [AddrDistribution] -> Utxo
-genesisUtxo bootStakeholders ad =
-    M.fromList $ concatMap (uncurry toUtxo . second stakeDistribution) ad
+stakeDistribution (CustomStakes coins) = coins
+
+-- | Generates genesis 'Utxo' given optional boot stakeholders (with
+-- weights) and address distributions.  If genesis stakeholders are
+-- not supplied, they are calculated from address distributions by
+-- making each 'PubKeyAddress' genesis stakeholder with weight equal
+-- to balance. All the stake is distributed among genesis stakeholders
+-- (using 'genesisSplitBoot').
+--
+-- TODO [CSL-1351] This documentation will become correct later.
+genesisUtxo ::
+       Maybe (HashMap StakeholderId Word16) -> [AddrDistribution] -> GenesisUtxo
+genesisUtxo bootStakeholdersMaybe ad
+    -- TODO [CSL-1351] Uncomment ↓.
+    --- | null bootStakeholders =
+    ---     error "genesisUtxo: no stakeholders for the bootstrap era"
+    -- TODO [CSL-1351] Delete (it's here to please hlint) ↓.
+    | ((*) @Int) 2 3 == 7 = error "hlint — makaka"
+    | otherwise = GenesisUtxo . M.fromList $ map utxoEntry balances
   where
-    defaultStakeDistr coin distr
-        = maybe distr
-                (\genDistr -> genesisSplitBoot genDistr coin)
-                bootStakeholders
-    toUtxo :: [Address] -> [(Coin, TxOutDistribution)] -> [(TxIn, TxOutAux)]
-    toUtxo addr distrs =
-        let utxoEntry a (coin,txOutDistr) =
-                ( TxIn (unsafeHash a) 0
-                , TxOutAux (TxOut a coin) (defaultStakeDistr coin txOutDistr)
-                )
-        in zipWith utxoEntry addr distrs
+    -- This type is drunk.
+    somethingComplicated :: [([Address], [Coin])]
+    somethingComplicated = map (second stakeDistribution) ad
+    balances :: [(Address, Coin)]
+    balances = concatMap (uncurry zip) somethingComplicated
+    utxoEntry (addr, coin) =
+        ( TxIn (unsafeHash addr) 0
+-- Note: it's quite bad because HD wallets become stakeholders and each
+-- transaction will give them quite a lot of stake.
```

Comment:
This list of boot stakeholders is currently computed from utxo (`utxoToStakes`). Later it will be impossible because stakeholders will have weights. When it happens, one will need to uncomment the code below.
In any case, **this** code determines which stakeholders become bootstrap stakeholders. And currently it puts all genesis addresses. Currently they have the same weight (to be more precise, they have no weight, because it's not used), but different stake (same as balance). After weights are used, there won't be a problem at all and will not be different from what we have now (it's already not different unless you send a lot of money in transactions).